pkg/scheduler/scheduler.go

```diff
@@ -149,14 +190,19 @@ func (s *Scheduler) schedule(ctx context.Context) {
 			//      evaluated in the same cycle is not taken into account.
 			//   2. An already admitted workload from a different cluster queue is borrowing, since all workloads
 			//      evaluated in the current cycle will compete for the resources that are not borrowed.
-			if usedCohorts.Has(cq.Cohort.Name) && (e.assignment.Borrows() || cq.Cohort.HasBorrowingQueues()) {
-				e.status = skipped
-				e.inadmissibleMsg = "other workloads in the cohort were prioritized"
-				continue
+			if cycleCohortsUsage.hasConflictWith(cq.Cohort.Name, e.assignment.Usage) && (e.assignment.Borrows() || cq.Cohort.HasBorrowingQueues()) {
+				// Check if the assignment still fits in the cohort.
+				sum := cycleCohortsUsage.getSumWith(cq.Cohort.Name, e.assignment.Usage)
```

Comment:
The [issue description](https://github.com/kubernetes-sigs/kueue/issues/1036) says `However, the check is too coarse: if the borrowing is happening in a different flavor, we shouldn't block it.`. This suggests we should only check if there is an intersection in flavors between  `cycleCohortsUsage` and `e.assignment`. 

Checking if the workloads CanFit again IMO adds complexity to the scheduler layer (I'm not sure it is correct yet), and it does not seem to be required by the issue description. 

Then we could follow up with another PR / Issue to improve if needed. wdyt @alculquicondor ?

Reply:
The PR is doing a bit more that the issue description.
To check if the flavor/resource is used in some kind of borrowing,  we need to keep track of them anyway,  since that was done also keeping track of quantity is fairly easy, and we avoid the potential need for the workload to do a full reassignment in a next scheduling cycle.

Reply:
It might be worth doing indeed, but I would consider a follow up PR, which might have a different priority, 

I think checking the CanFit at the scheduler layer adds complexity since the checks are already performed at the `flavorassigner` level and propagated up by `representativeMode`. One consequence (not a blocking one, but requiring being careful) is that the assignment failure messages are typically more descriptive in `flavorassigner`, and thus more actionable, than: "other workloads in the cohort were prioritized". 

Still, I'm just raising this as a potential place to split out the PRs, I will let @alculquicondor decide.

Reply:
>I think checking the CanFit at the scheduler layer adds complexity since the checks are already performed at the flavorassigner level and propagated up by representativeMode. One consequence (not a blocking one, but requiring being careful) is that the assignment failure messages are typically more descriptive in flavorassigner, and thus more actionable, than: "other workloads in the cohort were prioritized".

- The quota checks done in the flavorassigner has no knowledge of other workload from the same cohort that might be admitted in the same cycle.
 - If a workload admission fails in the flavorassigner it's failure status will be preserved. (it will not reach the cohort / borrowing check). "other workloads in the cohort were prioritized" will be set only if the assignment is successful in the queue but will not fit in the cohort. https://github.com/kubernetes-sigs/kueue/blob/7f2b00b796dd719e7add27c244278ce251f48dfe/pkg/scheduler/scheduler.go#L141-L143